### PR TITLE
New jsonrpc-core with futures and metadata support

### PR DESCRIPTION
**Pull request by tomusdrw**
- `jsonrpc-core` was rewritten to use `Futures` to handle async code.
- `Metadata` is now part of `jsonrpc-core`, so there is no need to do middlewares and augmentation

This will allow us to handle requests coming from different transports in a more elegant way (for instance `eth_accounts` can return all accounts for HTTP and IPC and only alter the list for dapps).

### CODE REVIEW

**Comment by gavofyork** [comment]
looks reasonable to me.

**Comment by gavofyork** [comment]
will need rebase...

**Comment by gavofyork** [comment]
still some conflicts...

File: rpc/src/v1/metadata.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	pub dapp_id: Option<String>,
+	/// Request origin
+	pub origin: Origin,
+}
+
+/// RPC request origin
+#[derive(Clone, Debug, PartialEq)]
+pub enum Origin {
+	/// RPC server
+	Rpc,
```
Comment (rphmeier) [inline rpc/src/v1/metadata.rs]: What does an origin of `Rpc` signify? All requests are RPC requests.

Comment (tomusdrw) [inline rpc/src/v1/metadata.rs]: It's JSON-RPC over HTTP coming from standalone JSON-RPC server (not dapps). It's initialized in crate `rpc` and also `parity/rpc.rs` so I thought that it makes sense to use this name to indentify it.

**Comment by tomusdrw** [comment]
IMHO it shouldn't go to `1.5.0` release, it doesn't really bring any new functionality and we need some more cross-platform testing.

**Comment by arkpar** [comment]
requires rebase

**Pull request merged**